### PR TITLE
Fix unstructured codegen

### PR DESCRIPTION
**Pull request by twicki**
## Technical Description

Fixes the codegen so both back ends are working again

### Example

Run the prototype examples

### Dependencies

This PR is independent

### CODE REVIEW

File: dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp [hunk 1]
```diff
@@ -73,7 +73,7 @@ void ASTStencilBody::visit(const std::shared_ptr<iir::ReductionOverNeighborExpr>
     }
   };
   std::string typeString = getLocationTypeString(expr->getRhsLocation());
-  ss_ << std::string(indent_, ' ') << "reduce" + typeString + "ToCell(libtag_t(), m_mesh, t, ";
+  ss_ << std::string(indent_, ' ') << "reduce" + typeString + "ToCell(libtag_t{}, m_mesh, t, ";
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp]: I think `libtag_t` is quite uncommon to be used as name for a template argument. Why not `LibTag` or something like that?

File: dawn/prototype/atlas_interface.hpp [hunk 2]
```diff
@@ -61,10 +61,17 @@ class Field {
 };
 
 template <typename T>
-Field<T> fieldType(atlasTag);
+Field<T> e_fieldType(atlasTag);
```
[the remark this answers is not in the document]
Comment (twicki) [reply dawn/prototype/atlas_interface.hpp]: done

File: dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp (hunk 1 above)
Comment (twicki) [reply dawn/src/dawn/CodeGen/CXXNaive-ico/ASTStencilBody.cpp]: done

File: dawn/prototype/mylibDriver.cpp [hunk 3]
```diff
@@ -8,8 +8,8 @@ using namespace mylibInterface;
 int main() {
   int w = 20;
   Mesh m{w, w, true};
-  CellField<double> out(m);
-  EdgeField<double> in(m);
+  mylib::FaceData<double> out(m);
```
[the remark this answers is not in the document]
Comment (twicki) [reply dawn/prototype/mylibDriver.cpp]: sure

File: dawn/prototype/atlas_interface.hpp [hunk 4]
```diff
@@ -87,11 +94,74 @@ std::vector<int> const& cellNeighboursOfCell(atlas::Mesh const& m, int const& id
...
 }
+template <typename Init, typename Op>
+auto reduceEdgeToCell(atlasTag, atlas::Mesh const& m, int idx, Init init, Op&& op) {
+  for(auto&& obj : edgeNeighboursOfCell(m, idx))
+    op(init, obj);
+  return init;
+}
+template <typename Init, typename Op>
+auto reduceVertexToCell(atlasTag, atlas::Mesh const& m, int idx, Init init, Op&& op) {
+  // for(auto&& obj : cellNeighboursOfCell(m, idx))
```
Comment (lukasm91) [inline dawn/prototype/atlas_interface.hpp]: can you put asserts here?

**Pull request merged**